From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Tue, 29 Aug 2023 22:29:08 +0200
Subject: [PATCH] Faster floating-point positive modulo

License: GPL-3.0-only (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

The JMH benchmark of this patch can be found in SunBox's `FasterFloatingPointPositiveModulo`

diff --git a/net/minecraft/util/Mth.java b/net/minecraft/util/Mth.java
index 47e6b1a772698d0172223a18748158e2e5381073..611a7305fca7852af541d95e1f91ac296bc82df2 100644
--- a/net/minecraft/util/Mth.java
+++ b/net/minecraft/util/Mth.java
@@ -167,14 +167,26 @@ public class Mth {
         return Math.floorMod(input, mod);
     }
 
-    public static float positiveModulo(final float input, final float mod) {
+    public static float positiveModuloForAnyDenominator(final float input, final float mod) {
         return (input % mod + mod) % mod;
     }
 
-    public static double positiveModulo(final double input, final double mod) {
+    public static double positiveModuloForAnyDenominator(final double input, final double mod) {
         return (input % mod + mod) % mod;
     }
 
+    // Gale start - faster floating-point positive modulo
+    public static float positiveModuloForPositiveIntegerDenominator(float input, float mod) {
+        var modulo = input % mod;
+        return modulo < 0 ? modulo + mod : modulo;
+    }
+
+    public static double positiveModuloForPositiveIntegerDenominator(double input, double mod) {
+        var modulo = input % mod;
+        return modulo < 0 ? modulo + mod : modulo;
+    }
+    // Gale end - faster floating-point positive modulo
+
     public static boolean isMultipleOf(final int dividend, final int divisor) {
         return dividend % divisor == 0;
     }
diff --git a/net/minecraft/world/level/levelgen/blending/Blender.java b/net/minecraft/world/level/levelgen/blending/Blender.java
index 6f03d4404c3fe5e36c236da294fc6e6561a965c4..b7722239afbc04f77445584f5019fedc3c8f3ced 100644
--- a/net/minecraft/world/level/levelgen/blending/Blender.java
+++ b/net/minecraft/world/level/levelgen/blending/Blender.java
@@ -188,7 +188,7 @@ public class Blender {
     private static double heightToOffset(final double height) {
         double dimensionFactor = 1.0;
         double targetY = height + 0.5;
-        double targetYMod = Mth.positiveModulo(targetY, 8.0);
+        double targetYMod = Mth.positiveModuloForPositiveIntegerDenominator(targetY, 8.0); // Gale - faster floating-point positive modulo
         return 1.0 * (32.0 * (targetY - 128.0) - 3.0 * (targetY - 120.0) * targetYMod + 3.0 * targetYMod * targetYMod) / (128.0 * (32.0 - 3.0 * targetYMod));
     }
 
